pkg/bloomcompactor/controller.go

```diff
@@ -211,19 +203,49 @@ func (s *SimpleBloomController) loadWorkForGap(
 	tenant string,
 	id tsdb.Identifier,
 	gap gapWithBlocks,
-) (v1.CloseableIterator[*v1.Series], []*bloomshipper.CloseableBlockQuerier, error) {
+) (v1.CloseableIterator[*v1.Series], v1.CloseableIterator[*bloomshipper.CloseableBlockQuerier], error) {
 	// load a series iterator for the gap
 	seriesItr, err := s.tsdbStore.LoadTSDB(ctx, table, tenant, id, gap.bounds)
 	if err != nil {
 		return nil, nil, errors.Wrap(err, "failed to load tsdb")
 	}
 
-	blocks, err := s.bloomStore.FetchBlocks(ctx, gap.blocks)
+	// load a blocks iterator for the gap
+	fetcher, err := s.bloomStore.Fetcher(table.ModelTime())
 	if err != nil {
-		return nil, nil, errors.Wrap(err, "failed to get blocks")
+		return nil, nil, errors.Wrap(err, "failed to get fetcher")
 	}
 
-	return seriesItr, blocks, nil
+	blocksIter, err := newBatchedBlockLoader(ctx, fetcher, gap.blocks)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to load blocks")
+	}
+
+	return seriesItr, blocksIter, nil
+}
+
+func (s *SimpleBloomController) closeLoadedBlocks(toClose []io.Closer, it v1.CloseableIterator[*bloomshipper.CloseableBlockQuerier]) {
+	// close loaded blocks
+	var err multierror.MultiError
+	for _, closer := range toClose {
+		err.Add(closer.Close())
+	}
+
+	switch itr := it.(type) {
+	case *batchedBlockLoader:
+		// close remaining loaded blocks from batch
+		err.Add(itr.CloseBatch())
+	default:
+		// close remaining loaded blocks
+		for itr.Next() && itr.Err() == nil {
```

Comment:
Would we ever reach here? As far as I can see `closeLoadedBlocks` is only called from `buildBlocks` which will pass whatever `loadWorkForGap` returns.

Reply:
Since the function accepts an interface `v1.CloseableInterator[*bloomshipper.CloseableBlockQuerier]` and not a concrete type, it could be possible that a non-batched version of the iterator is passed.